specs/protocol/isthmus/derivation.md

```diff
@@ -4,21 +4,48 @@
 **Table of Contents**
 
-- [Network upgrade automation transactions](#network-upgrade-automation-transactions)
-  - [L1Block deployment](#l1block-deployment)
-  - [GasPriceOracle deployment](#gaspriceoracle-deployment)
-  - [Operator fee vault deployment](#operator-fee-vault-deployment)
-  - [L1Block Proxy Update](#l1block-proxy-update)
-  - [GasPriceOracle Proxy Update](#gaspriceoracle-proxy-update)
-  - [OperatorFeeVault Proxy Update](#operatorfeevault-proxy-update)
-  - [GasPriceOracle Enable Isthmus](#gaspriceoracle-enable-isthmus)
-  - [EIP-2935 Contract Deployment](#eip-2935-contract-deployment)
+- [Network upgrade block rules](#network-upgrade-block-rules)
+  - [Upgrade Gas](#upgrade-gas)
+  - [Omission of user transactions](#omission-of-user-transactions)
+  - [Isthmus network upgrade automation transactions](#isthmus-network-upgrade-automation-transactions)
+    - [L1Block deployment](#l1block-deployment)
+    - [GasPriceOracle deployment](#gaspriceoracle-deployment)
+    - [Operator fee vault deployment](#operator-fee-vault-deployment)
+    - [L1Block Proxy Update](#l1block-proxy-update)
+    - [GasPriceOracle Proxy Update](#gaspriceoracle-proxy-update)
+    - [OperatorFeeVault Proxy Update](#operatorfeevault-proxy-update)
+    - [GasPriceOracle Enable Isthmus](#gaspriceoracle-enable-isthmus)
+    - [EIP-2935 Contract Deployment](#eip-2935-contract-deployment)
 - [Span Batch Updates](#span-batch-updates)
   - [Activation](#activation)
 
 <!-- END doctoc generated TOC please keep comment here to allow auto update -->
 
-# Network upgrade automation transactions
+# Network upgrade block rules
+
+## Upgrade Gas
+
+Starting with Isthmus, a change to the gas limit of upgrade blocks is introduced to ensure that
+upgrade transactions cannot run out of gas:
+If a hardfork activation block contains upgrade deposit transactions, the total sum of all upgrade
+transaction's gas limits, called *upgrade gas*, is added to the gas limit of the payload attributes.
+
+Before this change, it was possible that a chain with a low gas limit would not have enough gas to
+execute all upgrade transactions. Note that the default *system transaction max gas* is `1e6` and
+the only constraint on the *block gas limit* is that it has to be larger or equal to the *system
+transaction max gas* plus the *max resource limit* (default: `2e7`, may fully be consumed by user
+deposit transactions). So if the L1 deposit transaction plus upgrade transactions would consume more
+than the *system transaction max gas*, there was a risk of running out of gas.
+
+## Omission of user transactions
+
+Starting with Isthmus, network upgrade blocks must not contain any non-deposit transactions, to
+avoid any unintended interplay of user transactions with upgrade transactions.
+
+This is checked at the batch stage. So if a batch of an upgrade block is not empty, it is considered
+invalid and the rules for dropped batches apply.
```

Comment:
Why only omit non-deposit transactions? And why do we need both "upgrade gas" and "omission of user transactions"? It sounds like either of them alone would mitigate the issue

Reply:
You can't omit deposit transactions - they must be included as part of the censorship resistance.  Skipping user transactions removes the risk that the upgrade transactions don't use their full gas limit, leaving a higher than usual remaining block gas limit which could then be filled by particularly expensive L2 user transactions (ie calling precompiles that are expensive to execute in cannon). This way we know that the L1 deposit transactions have a fixed gas limit as usual (enforced on L1) and the potentially increased total gas limit is only usable by the upgrade transactions which we control so know they are fault provable.

Reply:
"Omission of user tx" only buys you the remainder of the block gas limit, which can be zero because the only guarantee is that `maxResourceLimit + systemTxMaxGas <= gasLimit` so it can be equal to the left side, which could in theory be fully exhausted.